### ISSUE TITLE
msal:loginFailure not triggering with msal.login.error

### ISSUE THREAD

**Issue by chr1soscl** [AzureAD/microsoft-authentication-library-for-js#1001]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[X] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Other... Please describe:
</code></pre>

## Browser:
- [X] Chrome version 75.0.3770.90

## Library version

<pre><code>
0.1.4
</code></pre>

## Current behavior
msal:loginFailure is not triggering, but I am getting a valid msal.login.error in the sessionStorage.

## Expected behavior
if msal.login.error occurs, msal:loginFailure should trigger.
<pre><code>
AADSTS50105: The signed in user '{EmailHidden}' is not assigned to a role for the application 'applicationId'(applicationName).
</code></pre>

## Minimal reproduction of the problem with instructions
Generate intentionally a scenario in which there is a login error, msal:loginFailure clause should execute, print something to console to verify. Note: login success works properly.

**Comment by sameerag** [comment]
@jasonnutter  one thing to keep in mind, @pkanher617 and Me are planning to streamline errors and may get rid of msal.login.error all together and just have one error json in cache. We should design the angular interface around the same.

**Comment by jasonnutter** [comment]
@chr1soscl Can you please provide more information on when this event isn't firing? For example, are you using `loginRedirect` or `loginPopup`? What are you doing that would cause a login error (e.g. closing the login popup)?

**Comment by chr1soscl** [comment]
> @chr1soscl Can you please provide more information on when this event isn't firing? For example, are you using `loginRedirect` or `loginPopup`? What are you doing that would cause a login error (e.g. closing the login popup)?

Hi Jason,
I am not calling loginRedirect or loginPopup, I am only setting the msalConfig with my clientId, authority, popup flag, consentScopes and redirectURL, when I try to open the app url it immediately redirects me to the Microsoft login screen, no need to call methods, we are testing with one user in the application group and one out of it, the one with access to the group is getting login successfully and the msal:loginSuccess is triggering, but for the user out of the group the msal:loginFailure is not triggering, but we can see in the sessionStorage the login error msal.login.error as provided above.

**Comment by jasonnutter** [comment]
@chr1soscl Ah okay, it sounds like you are using the MsalGuard, correct? We'll take a look into why that event might not be firing.

**Comment by chr1soscl** [comment]
> @chr1soscl Ah okay, it sounds like you are using the MsalGuard, correct? We'll take a look into why that event might not be firing.

yes, I am using the msalguard.

**Comment by chr1soscl** [comment]
> @chr1soscl Ah okay, it sounds like you are using the MsalGuard, correct? We'll take a look into why that event might not be firing.

@jasonnutter I have a question, currently we implemented our login to get authenticated and seems to be working mostly, but when we get a loginSuccess event triggered we are getting an accessToken and idToken under some strange key ({authority,clientId,scopes,userIdentified}), due to this strange Key I had to implement a different call to get a silent token to get the accessToken, why not to give back the accessToken inside a key with that name?

**Comment by chr1soscl** [comment]
> > @chr1soscl Ah okay, it sounds like you are using the MsalGuard, correct? We'll take a look into why that event might not be firing.
> 
> @jasonnutter I have a question, currently we implemented our login to get authenticated and seems to be working mostly, but when we get a loginSuccess event triggered we are getting an accessToken and idToken under some strange key ({authority,clientId,scopes,userIdentified}), due to this strange Key I had to implement a different call to get a silent token to get the accessToken (such a slow call, takes long to complete), why not to give back the accessToken inside a key with that name? I tried pulling it by concatenating this big key but not working, I could also retrieve it by index number, but if there are changes in the order in which the key are being returned, it will break the code.

**Comment by jasonnutter** [comment]
@chr1soscl Correct, keys are stored in the browser by a composite key made up of the authority, clientid, scopes, and user id in order to support storing many keys at once. You should not try to create this key yourself, as it is an implementation detail of `msal` and could change at any time (and you should certainly not use the index, which is not reliable). You should instead use the key as it returned to you in the broadcasted events, or use `acquireTokenSilent`.

**Comment by chr1soscl** [comment]
> @chr1soscl Correct, keys are stored in the browser by a composite key made up of the authority, clientid, scopes, and user id in order to support storing many keys at once. You should not try to create this key yourself, as it is an implementation detail of `msal` and could change at any time (and you should certainly not use the index, which is not reliable). You should instead use the key as it returned to you in the broadcasted events, or use `acquireTokenSilent`.

Hi @jasonnutter , what do you mean by the key returned by the broadcasted events? I cannot find the access token among the information returned by this.msalService.getUser(), I want to avoid use 'acquireTokenSilent' because it is a very slow call, it can take up to two seconds to complete.

**Comment by jasonnutter** [comment]
@chr1soscl Apologies, misspoke in my answer above. What I meant to say is that you should use the **token** as it is returned to you in the broadcasted events, or use `acquireTokenSilent`. 

`acquireTokenSilent` should be a fast call if it is able to retrieve the token from cache, so if its taking longer, its possible you aren't correctly retrieving it, resulting in a cache miss, which results in a network call. 

After you login with those consent scopes, how many access tokens do you see in local storage (with the composite key mentioned above)? Are you saying when you try to retrieve an access tokens using those scopes it doesn't work?

**Comment by chr1soscl** [comment]
> @chr1soscl Apologies, misspoke in my answer above. What I meant to say is that you should use the **token** as it is returned to you in the broadcasted events, or use `acquireTokenSilent`.
> 
> `acquireTokenSilent` should be a fast call if it is able to retrieve the token from cache, so if its taking longer, its possible you aren't correctly retrieving it, resulting in a cache miss, which results in a network call.
> 
> After you login with those consent scopes, how many access tokens do you see in local storage (with the composite key mentioned above)? Are you saying when you try to retrieve an access tokens using those scopes it doesn't work?

Hi @jasonnutter , the above line is part of my msalconfig, with only this configuration I only get one composite key, in the composite key as scope it is only mentioned my Angular clientId, the access token that comes with this composite key does not work with the api calls, I imagine because it is not considering the second parameter the api client id, if I add the call to acquireTokenSilent, only then I get another composite key with the proper consent scope and the accessToken provided works fine with the api.

**Comment by andrantis** [comment]
I'm experience a similar issue. I'm using `loginRedirect()` and I get this in the console `Redirect Success:  {uniqueId: "", tenantId: "", tokenType: "", idToken: null, idTokenClaims: null, …}`

The scenario is when a user is not assigned to the app. I would expect that instead of the select user screen you get a screen saying the user doesn't have permission to access the app. Instead I get redirected back to the app with the above. 
There is also a `msal.error: interaction_required` and `msal.login.error: AADSTS50105: The signed in user '{EmailHidden}' is not assigned to a role for the application xxx` stored in localstorage.

Is there a different way to handle the case for users that are not assigned to the app?

By the way, I'm using the current sample app for angular 8.

**Comment by jasonnutter** [comment]
@andrantis That doesn't seem right, can you debug `handleRedirectAuthenticationResponse` in `UserAgentApplication` and see why MSAL might think the response was successful?

**Comment by andrantis** [comment]
@jasonnutter looks like `handleRedirectCallback` gets an `authResponse` with only `state` in it. That causes to broadcast an `msal:acquireTokenSuccess` (I don't think this is right) and also execute `authOrTokenCallback` with a `null` as error and the `authResponse`.

**Comment by jasonnutter** [comment]
@andrantis Interesting. Can you try `1.3.0-beta.1` and see if it still occurs? We made some fixes to processing responses.

**Comment by andrantis** [comment]
@jasonnutter Yes it still the same with `1.3.0-beta.1`. Am I setting up something wrong or my flow is wrong? I'm doing `acquireTokenSilent()` and if that fails then do `loginRedirect()`. It works fine if the user or group is assigned to the the enterprise application. But if the user/group is not part of the enterprise application, the only way I found to stop the infinite redirect loop is to look for an `authError === null && response.IdToken === null`, otherwise it keeps looping without prompting the user to interact. It's weird because the `msal.error` in Session Storage does say `interaction_required`. Do you know if there is a specific option or setting I need to set?

**Comment by jasonnutter** [comment]
@andrantis Ah, are you invoking `loginRedirect` for all `acquireTokenSilent` errors? You should be checking the error code and only invoking one of the interactive methods if the error is a user interaction error. You should not be invoking `loginRedirect` for an error (such as group assignment) that is only resolvable by admins.

```js
import { UserAgentApplication, InteractionRequiredAuthError } from "msal";
 
msal.acquireTokenSilent()
  .catch(error => {
     if (InteractionRequiredAuthError.isInteractionRequiredError(error.errorCode)) {
        msal.acquireTokenRedirect();
     } else {
        // show an error to the user
     }
  });

**Comment by jasonnutter** [comment]
@andrantis We found the bug for your issue, which was fixed in #1562. This will be available in the next version of MSAL Angular.
